[v4] indent: false positive with multi-line async function call + destructuring assignment

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

    1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
    3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** 4.1.1
* **Node Version:** 4.3.1
* **npm Version:** 2.14.12

**What parser (default, Babel-ESLint, etc.) are you using?**

babel-eslint 7.2.3

**Please show your full configuration:**

<!-- Paste your configuration below: -->
```
{
    "env": {
        "node": true,
        "es6": true,
    },
    "parser": "babel-eslint",
    "parserOptions": {
        "ecmaFeatures": {
            "arrowFunctions": true,
            "forOf": true,
            "defaultParams": true,
            "destructuring": true,
            "objectLiteralShorthandProperties": true,
            "restParams": true,
            "spread": true,
        },
    },
    "rules": {
        "indent": ["warn", 4, {"SwitchCase": 1}],
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

<!-- Paste the source code below: -->
```js
async function test() {
    const {
        foo,
        bar,
    } = await doSomethingAsync(
        1,
        2,
        3,
    );
}
```

**What did you expect to happen?**

No errors

**What actually happened? Please include the actual, raw output from ESLint.**
```
  6:1  warning  Expected indentation of 12 spaces but found 8  indent
  7:1  warning  Expected indentation of 12 spaces but found 8  indent
  8:1  warning  Expected indentation of 12 spaces but found 8  indent
  9:1  warning  Expected indentation of 8 spaces but found 4   indent
```

NOTES:
- It works if I remove the `async` keyword
- It works if I don't use destructuring assignment
- It works if I don't split the function arguments into multiple lines

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -887,9 +887,7 @@ module.exports = {
         function addFunctionCallIndent(node) {
             let openingParen;
 
-            if (node.parent.type === "AwaitExpression") {
-                openingParen = node.parent;
-            } else if (node.arguments.length) {
+            if (node.arguments.length) {
                 openingParen = sourceCode.getFirstTokenBetween(node.callee, node.arguments[0], astUtils.isOpeningParenToken);
             } else {
                 openingParen = sourceCode.getLastToken(node, 1);

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -887,7 +887,9 @@ module.exports = {
         function addFunctionCallIndent(node) {
             let openingParen;
 
-            if (node.arguments.length) {
+            if (node.parent.type === "AwaitExpression") {
+                openingParen = node.parent;
+            } else if (node.arguments.length) {
                 openingParen = sourceCode.getFirstTokenBetween(node.callee, node.arguments[0], astUtils.isOpeningParenToken);
             } else {
                 openingParen = sourceCode.getLastToken(node, 1);

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -1032,7 +1032,7 @@ module.exports = {
                 const nodeTokens = getTokensAndComments(node);
                 const tokensFromOperator = nodeTokens.slice(lodash.sortedIndexBy(nodeTokens, operator, token => token.range[0]));
 
-                offsets.setDesiredOffsets(tokensFromOperator, sourceCode.getFirstToken(node.left), 1);
+                offsets.setDesiredOffsets(tokensFromOperator, sourceCode.getLastToken(node.left), 1);
                 offsets.ignoreToken(tokensFromOperator[0]);
                 offsets.ignoreToken(tokensFromOperator[1]);
             },
@@ -1327,12 +1327,12 @@ module.exports = {
             VariableDeclarator(node) {
                 if (node.init) {
                     const equalOperator = sourceCode.getTokenBefore(node.init, astUtils.isNotOpeningParenToken);
-                    const tokenAfterOperator = sourceCode.getTokenAfter(equalOperator);
+                    const tokensAfterOperator = sourceCode.getTokensAfter(equalOperator, token => token.range[1] <= node.range[1]);
 
                     offsets.ignoreToken(equalOperator);
-                    offsets.ignoreToken(tokenAfterOperator);
-                    offsets.matchIndentOf(equalOperator, tokenAfterOperator);
-                    offsets.matchIndentOf(sourceCode.getFirstToken(node), equalOperator);
+                    offsets.ignoreToken(tokensAfterOperator[0]);
+                    offsets.setDesiredOffsets(tokensAfterOperator, equalOperator, 1);
+                    offsets.matchIndentOf(sourceCode.getLastToken(node.id), equalOperator);
                 }
             },
 

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -3910,6 +3910,51 @@ ruleTester.run("indent", rule, {
                 }
             `
         },
+        {
+            code: unIndent`
+                function* test() {
+                    const {
+                        foo,
+                        bar,
+                    } = yield doSomethingAsync(
+                        1,
+                        2,
+                        3,
+                    );
+                }
+            `
+        },
+        {
+            code: unIndent`
+                ({
+                    a: b
+                } = +foo(
+                    bar
+                ));
+            `
+        },
+        {
+            code: unIndent`
+                const {
+                    foo,
+                    bar,
+                } = typeof foo(
+                    1,
+                    2,
+                    3,
+                );
+            `
+        },
+        {
+            code: unIndent`
+                const {
+                    foo,
+                    bar,
+                } = +(
+                    foo
+                );
+            `
+        },
 
         //----------------------------------------------------------------------
         // JSX tests

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -67,7 +67,7 @@ function unIndent(strings) {
     return lines.map(line => line.slice(minLineIndent)).join("\n");
 }
 
-const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } } });
+const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 8, ecmaFeatures: { jsx: true } } });
 
 ruleTester.run("indent", rule, {
     valid: [
@@ -3892,6 +3892,22 @@ ruleTester.run("indent", rule, {
                         ok: false
                     }
                 )
+            `
+        },
+
+        // https://github.com/eslint/eslint/issues/8815
+        {
+            code: unIndent`
+                async function test() {
+                    const {
+                        foo,
+                        bar,
+                    } = await doSomethingAsync(
+                        1,
+                        2,
+                        3,
+                    );
+                }
             `
         },
 

